### PR TITLE
Support asm/cmm/js sources in executable components (#8639)

### PR DESCRIPTION
**Pull request by JoshMeredith**
Implementation for https://github.com/haskell/cabal/issues/8639

The code for building each of `c`/`cpp`/`asm`/`cmm`/`js` is factorised into a local function `buildExtraSources`. The local function is separate per library/executable building - i.e. no factorisation between these two stages is attempted.

Added tests:
- `cabal-testsuite/PackageTests/CmmSourcesExe`: based on the sibling `CmmSources` test, with the source directories and `.cmm` 
file moved into the executable, and the library removed
- `cabal-testsuite/PackageTests/JS/JsSourcesExe`: as above for the `.js` source.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

### CODE REVIEW

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 1]
```diff
@@ -808,213 +808,75 @@ buildOrReplLib mReplFlags verbosity numJobs pkg_descr lbi lib clbi = do
                 else do vanilla; shared
       whenProfLib (runGhcProg profOpts)
 
+  let
+    buildExtraSources mkSrcOpts wantDyn srcs =
+      sequence_
+        [ do
```
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:814]: I find the `sequence_` + list-comprehension odd. Could we use:

```suggestion
    buildExtraSources mkSrcOpts wantDyn srcs = traverse_ (buildExtraSource mkSrcOpts wantDyn) srcs
    buildExtraSource  mkSrcOpts wantDyn filename = do
```

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 2]
```diff
@@ -1876,95 +1761,76 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
         , ghcOptNumJobs = numJobs
         }
 
+  let
+    buildExtraSources mkSrcOpts wantDyn srcs =
+      sequence_
```
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:1766]: Same here

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 3]
```diff
@@ -1876,95 +1761,76 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
...
-        createDirectoryIfMissingVerbose verbosity True odir
-        needsRecomp <- checkNeedsRecompilation filename opts
-        when needsRecomp $
-          runGhcProg opts
-      | filename <- cSrcs
-      ]
+    buildExtraSources Internal.componentCcGhcOptions True cSrcs
+
+  -- build any JS sources
+  unless (not hasJsSupport || null (jsSourceFiles buildSources)) $ do
```
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:1821]:
```suggestion
  unless (not hasJsSupport || null jsSrcs) $ do
```

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 4]
```diff
@@ -1876,95 +1761,76 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
...
+    info verbosity "Building JS Sources..."
+    buildExtraSources Internal.componentJsGhcOptions False jsSrcs
+
+  -- build any ASM sources
+  unless (null $ asmSourceFiles buildSources) $ do
```
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:1826]:
```suggestion
  unless (null asmSrcs) $ do
```

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 5]
```diff
@@ -1876,95 +1761,76 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
...
+    info verbosity "Building Assembler Sources..."
+    buildExtraSources Internal.componentAsmGhcOptions True asmSrcs
+
+  -- build any Cmm sources
+  unless (null $ cmmSourceFiles buildSources) $ do
```
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:1831]:
```suggestion
  unless (null cmmSrcs) $ do
```

File: Cabal/src/Distribution/Simple/GHC.hs (hunk 1 above)
Comment (JoshMeredith) [reply Cabal/src/Distribution/Simple/GHC.hs]: Ah yes that is more readable than the list comprehension

**Review by hsyl20 — APPROVED** [review]
LGTM

**Comment by hsyl20** [comment]
Could someone review this PR, please? It would be nice to get it backported to 3.10 to avoid this surprising bug for users trying to use the JS backend in 9.8. Thanks!

**Review by Kleidukos — APPROVED** [review]

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.10

**Comment by hsyl20** [comment]
Thanks @Kleidukos !

**Comment by ulysses4ever** [comment]
@Mergify rebase

**Comment by ulysses4ever** [comment]
Fourmolu is failing for an unrelated reason. A fix is in flight.

**Comment by Mikolaj** [comment]
@Kleidukos, @hsyl20, @JoshMeredith: It turns out nobody resolved the conflicts in the backport https://github.com/haskell/cabal/pull/9130 and so this is not slated for release in cabal 3.10.2.0, but only in cabal 3.12.1.0 (which we optimistically expect to get released just after 3.10.2.0). Is that a problem? Is there a volunteer for an emergency conflict resolution to squeeze this in into 3.10 one minute before midnight?

**Comment by hsyl20** [comment]
It shouldn't be a problem, we'll just recommend 3.12 for the JS backend.

**Comment by alt-romes** [comment]
@JoshMeredith 

In this change, why does dynLinkerOpts not have jsObjs? I suppose we cannot dynamically link jsObjects?

```diff
@@ -1794,14 +1676,14 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
                 PD.extraFrameworkDirs bnfo
           , ghcOptInputFiles =
               toNubListR
-                [tmpDir </> x | x <- cLikeObjs ++ cxxObjs]
+                [tmpDir </> x | x <- cLikeObjs ++ cxxObjs ++ jsObjs ++ cmmObjs ++ asmObjs]
           }
       dynLinkerOpts =
         mempty
           { ghcOptRPaths = rpaths
           , ghcOptInputFiles =
               toNubListR
-                [tmpDir </> x | x <- cLikeObjs ++ cxxObjs]
+                [tmpDir </> x | x <- cLikeObjs ++ cxxObjs ++ cmmObjs ++ asmObjs]
           }
       replOpts =
         baseOpts
```

**Comment by Mikolaj** [comment]
@JoshMeredith: a humble pint?

@hsyl20: do you perhaps know the answer off the top of you head?

**Comment by JoshMeredith** [comment]
Hi, sorry I saw this and forgot to send the reply.

Yes, it's intentional that we don't currently dynamically link JavaScript. Our team hasn't discussed if it's possible, maybe @hsyl20 or @luite have ideas.

**Pull request merged**